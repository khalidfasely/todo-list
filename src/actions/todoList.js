import moment from 'moment';
import database from '../firebase/firebase';

// Add Todo Item
export const addTodoItem = (todoItem) => ({
    type: 'ADD_TODO_ITEM',
    todoItem
});

export const startAddTodoItem = (todoItemData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            content = '',
            active = false,
            time = moment(),
            timeOfDeactive = 0
        } = todoItemData;
        const todoItem = { content, active, time, timeOfDeactive };

        return database.ref(`users/${uid}/list`).push(todoItem).then((ref) => {
            dispatch(addTodoItem({
                id: ref.key,
                ...todoItem
            }));
        });
    };
};

// Deactive Item List
export const deactiveTodoItem = (id, updates) => ({
    type: 'DEACTIVE_TODO_ITEM',
    id,
    updates
});

export const startDeactiveTodoItem = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const updates = {
            active: false,
            timeOfDeactive: moment().format()
        };

        return database.ref(`users/${uid}/list/${id}`).update(updates).then(() => {
            dispatch(deactiveTodoItem(id, updates));
        });
    };
};

// Remove Item from List
export const removeTodoItem = (id) => ({
    type: 'REMOVE_TODO_ITEM',
    id
});

export const startRemoveTodoItem = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        
        return database.ref(`users/${uid}/list/${id}`).remove().then(() => {
            dispatch(removeTodoItem(id));
        });
    };
};

//Set Todo List
export const setTodoList = (todoList) => ({
    type: 'SET_TODO_LIST',
    todoList
});

export const startSetTodoList = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/list`)
            .once('value')
            .then((snapshot) => {
                const todoListData = [];

                snapshot.forEach((childSnapshot) => {
                    todoListData.push({
                      id: childSnapshot.key,
                      ...childSnapshot.val()
                    });
                });
                dispatch(setTodoList(todoListData));
            });
    };
};