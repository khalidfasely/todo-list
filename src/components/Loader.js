const Loader = () => (
    <div className="loader-page">
        <div className="loader-container">
            <div className="loader">
                <h2 data-text='Loading...'>Loading...</h2>
            </div>    
        </div>
    </div>
);

export default Loader;