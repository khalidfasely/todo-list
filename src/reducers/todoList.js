// TodoList Reducer

const todoListReducerDefaultState = [];

export default (state = todoListReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_TODO_ITEM':
            return [
                ...state,
                action.todoItem
            ];
        case 'DEACTIVE_TODO_ITEM':
            return state.map(item => {
                if (item.id === action.id) {
                    return {
                        ...item,
                        ...action.updates
                    }
                } else {
                    return item;
                }
            });
        case 'REMOVE_TODO_ITEM':
            return state.filter(({ id }) => id !== action.id);
        case 'SET_TODO_LIST':
            return action.todoList;
        default:
            return state;
    }
};