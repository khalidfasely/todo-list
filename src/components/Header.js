import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => {
    return (
        <header className="header">
            <div className="header__logout-button-container">
                <button
                    className="header__logout-button"
                    data-testid='logout-button'
                    onClick={startLogout}
                >
                    Logout
                </button>
            </div>
            <h1 className="header__title">
                Todo List
            </h1>
        </header>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);