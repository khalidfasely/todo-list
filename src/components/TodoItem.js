import moment from "moment"
import { connect } from "react-redux";
import { startDeactiveTodoItem, startRemoveTodoItem } from "../actions/todoList";

export const TodoItem = ({
        id,
        content,
        time,
        active,
        timeOfDeactive,
        startDeactiveTodoItem,
        startRemoveTodoItem
    }) => {
    
    const onDeactiveTodoItem = () => {
        startDeactiveTodoItem(id);
    };

    const onRemoveTodoItem = () => {
        startRemoveTodoItem(id);
    };

    return (
        <div>
            <div>{content}</div>
            <div>{moment(time).fromNow()}</div>
            {
                active ?
                <button data-testid='done-button' onClick={onDeactiveTodoItem}>Done</button> :
                <abbr title={moment(timeOfDeactive).fromNow()}>
                    <button disabled={true}>Already Done</button>
                </abbr>
            }
            <button data-testid='x-button' onClick={onRemoveTodoItem}>X</button>
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startRemoveTodoItem: (id) => dispatch(startRemoveTodoItem(id)),
    startDeactiveTodoItem: (id) => dispatch(startDeactiveTodoItem(id))
});

export default connect(undefined, mapDispatchToProps)(TodoItem);