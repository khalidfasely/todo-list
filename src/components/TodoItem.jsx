import moment from "moment"
import { connect } from "react-redux";
import { startDeactiveTodoItem, startRemoveTodoItem } from "../actions/todoList";

export const TodoItem = ({
        id,
        content,
        time,
        active,
        timeOfDeactive,
        startDeactiveTodoItem,
        startRemoveTodoItem
    }) => {
    
    const onDeactiveTodoItem = () => {
        startDeactiveTodoItem(id);
    };

    const onRemoveTodoItem = () => {
        startRemoveTodoItem(id);
    };

    return (
        <div className="item-container">
            <h4 className="item__content">{content}</h4>
            <div className="item__right">
                <div className="item__time">{moment(time).fromNow()}</div>
                {
                    active ?
                    <abbr title="Already done?">
                        <button className="item__button deactive" data-testid='done-button' onClick={onDeactiveTodoItem}>&#x2713;</button>
                    </abbr> :
                    <abbr title={`Done ${moment(timeOfDeactive).fromNow()}`}>
                        <button className="item__button deactivated" disabled={true}>&#x2713;</button>
                    </abbr>
                }
                <abbr title="Remove From List">
                    <button className="item__button remove" data-testid='x-button' onClick={onRemoveTodoItem}>X</button>
                </abbr>
            </div>
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startRemoveTodoItem: (id) => dispatch(startRemoveTodoItem(id)),
    startDeactiveTodoItem: (id) => dispatch(startDeactiveTodoItem(id))
});

export default connect(undefined, mapDispatchToProps)(TodoItem);