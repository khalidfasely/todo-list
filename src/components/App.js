import Header from "./Header";
import NewItem from "./NewItem";
import '../firebase/firebase';
import { useEffect } from "react";
import database from "../firebase/firebase";

function App() {
  //useEffect(() => {
  //  database.ref('users')
  //  .once('value')
  //  .then((snapshot) => {
  //    const todoListData = [];
  //  
  //    snapshot.forEach((childSnapshot) => {
  //      todoListData.push({
  //        //id: childSnapshot.key,
  //        ...childSnapshot.val()
  //      });
  //    });
  //  
  //    console.log(todoListData);
  //  });
  //}, [])
  return (
    <div className="App">
      <Header />
      <NewItem />
    </div>
  );
};

export default App;
