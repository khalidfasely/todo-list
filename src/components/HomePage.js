import NewItem from "./NewItem";
import Header from "./Header";
import TodoList from "./TodoList";

const HomePage = () => {
    return (
        <div className="app">
            <div>
                <Header />
                <div className="sections-container">
                    <div className="content-container">
                        <NewItem />
                        <TodoList />
                    </div>
                </div>
            </div>
        </div>
    )
};

export default HomePage;