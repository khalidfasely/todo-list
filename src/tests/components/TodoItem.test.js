import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { TodoItem } from '../../components/TodoItem';
import todoList from '../fixtures/todoList';

let props;

beforeEach(() => {
    props = {
        ...todoList[0],
        startDeactiveTodoItem: jest.fn(),
        startRemoveTodoItem: jest.fn()
    }
});

test('Should render TodoItem component correctly with active item', () => {
    const { asFragment } = render(<TodoItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render TodoItem component correctly with unactive item', () => {
    const { asFragment } = render(<TodoItem {...{...props, ...todoList[2]}} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should call onDeactiveTodoItem when Done button click', () => {
    const { getByTestId } = render(<TodoItem {...props} />);
    const doneButtonEl = getByTestId('done-button');
    fireEvent.click(doneButtonEl);
    expect(props.startDeactiveTodoItem).toHaveBeenCalled();
});

test('Should call onRemoveTodoItem when X button click', () => {
    const { getByTestId } = render(<TodoItem {...props} />);
    const xButtonEl = getByTestId('x-button');
    fireEvent.click(xButtonEl);
    expect(props.startRemoveTodoItem).toHaveBeenCalled();
});