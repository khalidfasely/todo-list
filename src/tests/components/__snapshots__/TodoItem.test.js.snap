// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render TodoItem component correctly with active item 1`] = `
<DocumentFragment>
  <div
    class="item-container"
  >
    <h4
      class="item__content"
    >
      todo1
    </h4>
    <div
      class="item__right"
    >
      <div
        class="item__time"
      >
        52 years ago
      </div>
      <abbr
        title="Already done?"
      >
        <button
          class="item__button deactive"
          data-testid="done-button"
        >
          ✓
        </button>
      </abbr>
      <abbr
        title="Remove From List"
      >
        <button
          class="item__button remove"
          data-testid="x-button"
        >
          X
        </button>
      </abbr>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render TodoItem component correctly with unactive item 1`] = `
<DocumentFragment>
  <div
    class="item-container"
  >
    <h4
      class="item__content"
    >
      todo3
    </h4>
    <div
      class="item__right"
    >
      <div
        class="item__time"
      >
        52 years ago
      </div>
      <abbr
        title="Done 52 years ago"
      >
        <button
          class="item__button deactivated"
          disabled=""
        >
          ✓
        </button>
      </abbr>
      <abbr
        title="Remove From List"
      >
        <button
          class="item__button remove"
          data-testid="x-button"
        >
          X
        </button>
      </abbr>
    </div>
  </div>
</DocumentFragment>
`;
